Add status probability and delay optional argument to the echo server

ldemailly opened the pull request:
http://localhost:8080/echo/?delay=1ms&status=567:33,568:33
For instance delays the response by 1ms and has 33% chance of returning
567, 33% 568 and 34% 200

Fixes #50

ldemailly committed:
Adding ?delay=<duration> to echo url

Addresses #50 partially

ldemailly committed:
Support status param as list with probability

http://localhost:8080/echo/?delay=1ms&status=567:33,568:33
For instance delays the response by 1ms and has 33% chance of returning
567, 33% 568 and 34% 200

Fixes #50

ldemailly committed:
Lint fix

ldemailly committed:
Adding tests for parseStatus

ldemailly committed:
Minor coverage increase

douglas-reid commented on fhttp/http.go:
```diff
@@ -750,9 +752,81 @@ var (
+	if len(lst) == 1 && !strings.ContainsRune(status, ':') {
+		s, err := strconv.Atoi(status)
+		if err != nil {
+			log.Warnf("Bad input status %v, not a number nor coma and colon separated % list", status)
```
nit: s/coma/comma

ldemailly committed:
Fixing missing %% in string to get %

douglas-reid commented on fhttp/http.go:
```diff
@@ -750,9 +752,81 @@ var (
+			log.Warnf("Percentage is not a [0 - 100] number in %v -> %v : %v %d", status, percStr, err, p)
+			return http.StatusBadRequest
+		}
+		if lastPercent+p > 100 {
```
[thinking out loud] this seems fine. i wonder if it might be nice to autoscale to 100 down-the-road. like if say:

`?status:404:95,302:15,200:90`

we would adjust down to something like:

```
404 -> 48
302 -> 7
200 -> 45
```

That might be a bit surprising, but might help correct for silly math blunders too...

douglas-reid commented on fhttp/http.go:
```diff
@@ -750,9 +752,81 @@ var (
+	if len(lst) == 1 && !strings.ContainsRune(status, ':') {
+		s, err := strconv.Atoi(status)
+		if err != nil {
+			log.Warnf("Bad input status %v, not a number nor coma and colon separated %% list", status)
```
nit: s/coma/comma

douglas-reid commented on fhttp/http.go:
```diff
@@ -750,9 +752,81 @@ var (
+		lastPercent += p
+	}
+	res := rand.Intn(100)
+	log.Debugf("[0-100[ for %s roll got %d -> %d", status, res, probTable[res])
```
do we need a full 100 entry array for this?

it seems like we could track this with a map of `len(lst)` with additive indices. we expect the number of status entries to be small and so running through the list each time is probably efficient enough.  this would also allow arbitrary precision on the percentages.

something like:
```
// for ?status=503:20,404:10

table := map[int]int{ 20:503, 30:404 }
res := rand.Intn(100)
for percent, code := range table {
  if res < percent {
     return code
  }
}
return http.StatusOK
```

not sure this matters all that much, but thought I'd ask.

douglas-reid replied:
or, maybe we can do this as we parse status codes and avoid any secondary storage altogether?  generate the number first, walk through as we go... (haven't thought through all the error conditions in this approach, however).

douglas-reid commented on fhttp/http_test.go:
```diff
@@ -248,6 +248,57 @@ func TestDebugSummary(t *testing.T) {
+	return int(float64(x)+50.) / 100
+}
+
+func TestParseStatus2(t *testing.T) {
```
nit: maybe call TestStatusDistribution ?  that would help immediately clarify the intent.

ldemailly committed:
Code review part1: the easy changes

Thx Doug!, working on the other next

ldemailly committed:
Code review suggestion part 2

Now implemented overkill floating point resolution probabilities :-)
No more 100 array, replaced by 2 arrays of the input len

ldemailly committed:
Better tests and rounding

ldemailly committed:
More rounding to actually test it's required

Only required for 4 digits so maybe float32() cast would be enough for
most cases

douglas-reid reviewed (approved):
/lgtm

overkill is the best kill.

ldemailly committed:
Last round of comment fixes

ldemailly committed:
Calculate % for each retcode and allow initial errors (#57)

Also handle 100% error case in graph title
